src/modules/MagneticFieldReader/MagneticFieldReaderModule.cpp

```diff
@@ -62,5 +62,87 @@ void MagneticFieldReaderModule::initialize() {
         }
         LOG(INFO) << "Set constant magnetic field: " << Units::display(b_field, {"T", "mT"});
+    } else if(field_model == MagneticField::MESH) {
+        type = MagneticFieldType::CUSTOM;
+
+        auto fallback_field = config_.get<ROOT::Math::XYZVector>("magnetic_field_fallback", ROOT::Math::XYZVector());
+        auto field_data = read_field();
+
+        auto dims = field_data.getDimensionality(); // Vector dimensionality of the field (should be 3 for B-field)
+        if(dims != 3) {
+            throw std::invalid_argument("B-field vectors were not three dimensional.");
+        }
+
+        auto cellsize = field_data.getSize();     // Physical dimensions for each cell in the field mesh
+        auto ncells = field_data.getDimensions(); // Number of cells in each direction of the field mesh
+        auto field_mesh = field_data.getData();   // The field mesh B-field data as a flattened array
+
+        if(ncells[0] * ncells[1] * ncells[2] * dims != field_mesh->size()) {
+            throw std::invalid_argument("B-field does not span the given dimensions");
+        }
+
+        MagneticFieldFunction function =
+            [fallback_field, field_mesh, cellsize, ncells, dims](const ROOT::Math::XYZPoint& coord) {
+                // Find the nearest field mesh cell to the given input coordinate (assuming the mesh is centred about the
+                // origin)
+                auto xind = std::round((coord.X() / cellsize[0]) + (static_cast<double>(ncells[0]) / 2.));
+                auto yind = std::round((coord.Y() / cellsize[1]) + (static_cast<double>(ncells[1]) / 2.));
+                auto zind = std::round((coord.Z() / cellsize[2]) + (static_cast<double>(ncells[2]) / 2.));
+
+                if(xind < 0 || xind >= static_cast<double>(ncells[0]) || yind < 0 ||
+                   yind >= static_cast<double>(ncells[1]) || zind < 0 || zind >= static_cast<double>(ncells[2])) {
+                    return fallback_field;
+                } else {
+                    auto field = field_mesh.get();
+
+                    long unsigned int ix = static_cast<long unsigned int>(xind);
+                    long unsigned int iy = static_cast<long unsigned int>(yind);
+                    long unsigned int iz = static_cast<long unsigned int>(zind);
+
+                    auto bfieldx = field->at(ix * ncells[1] * ncells[2] * dims + iy * ncells[2] * dims + iz * dims);
+                    auto bfieldy = field->at(ix * ncells[1] * ncells[2] * dims + iy * ncells[2] * dims + iz * dims + 1);
+                    auto bfieldz = field->at(ix * ncells[1] * ncells[2] * dims + iy * ncells[2] * dims + iz * dims + 2);
+                    return ROOT::Math::XYZVector(bfieldx, bfieldy, bfieldz);
+                }
+            };
+
+        geometryManager_->setMagneticFieldFunction(function, type);
+        auto detectors = geometryManager_->getDetectors();
+        for(auto& detector : detectors) {
```

Comment:
Since this loop appears twice now I think we can move it out of the `if`-`else if` block.